ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java

```diff
@@ -165,45 +191,56 @@ static boolean isValidBlobRecordVersion(short blobRecordVersion) {
    */
   public static BlobAll deserializeBlobAll(InputStream stream, StoreKeyFactory storeKeyFactory)
       throws IOException, MessageFormatException {
-    validateHeader(stream);
-    StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(stream));
-    BlobProperties blobProperties = deserializeBlobProperties(stream);
-    byte[] userMetadata = deserializeUserMetadata(stream).array();
-    BlobData blobData = deserializeBlob(stream);
-    return new BlobAll(storeKey, new BlobInfo(blobProperties, userMetadata), blobData);
-  }
-
-  /**
-   * Read and validate the message header from a complete blob record.
-   * @param stream the {@link InputStream} from which to read the blob record.
-   * @throws IOException
-   * @throws MessageFormatException
-   */
-  private static void validateHeader(InputStream stream) throws IOException, MessageFormatException {
+    boolean hasEncryptionKeyRecord = false;
     DataInputStream inputStream = new DataInputStream(stream);
     short headerVersion = inputStream.readShort();
+    ByteBuffer headerBuf;
     switch (headerVersion) {
       case Message_Header_Version_V1:
-        ByteBuffer headerBuf = ByteBuffer.allocate(MessageFormatRecord.MessageHeader_Format_V1.getHeaderSize());
+        headerBuf = ByteBuffer.allocate(MessageFormatRecord.MessageHeader_Format_V1.getHeaderSize());
         headerBuf.putShort(headerVersion);
         inputStream.read(headerBuf.array(), Version_Field_Size_In_Bytes,
             MessageHeader_Format_V1.getHeaderSize() - Version_Field_Size_In_Bytes);
         headerBuf.rewind();
         new MessageHeader_Format_V1(headerBuf).verifyHeader();
         break;
+      case Message_Header_Version_V2:
+        headerBuf = ByteBuffer.allocate(MessageFormatRecord.MessageHeader_Format_V2.getHeaderSize());
+        headerBuf.putShort(headerVersion);
+        inputStream.read(headerBuf.array(), Version_Field_Size_In_Bytes,
+            MessageHeader_Format_V2.getHeaderSize() - Version_Field_Size_In_Bytes);
+        headerBuf.rewind();
+        MessageHeader_Format_V2 header = new MessageHeader_Format_V2(headerBuf);
+        header.verifyHeader();
+        hasEncryptionKeyRecord =
+            header.getBlobEncryptionKeyRecordRelativeOffset() != Message_Header_Invalid_Relative_Offset;
+        break;
       default:
         throw new MessageFormatException("Message header version not supported",
             MessageFormatErrorCodes.Unknown_Format_Version);
     }
+    StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(stream));
+    BlobProperties blobProperties = deserializeBlobProperties(stream);
```

Comment:
I guess first comes the keyRecord and then the BlobProperties. Wondering do we have tests to cover this section.

Reply:
Yes - thanks for catching this. An earlier design had the encryption key record after the blob properties, I missed changing this.
The reason this didn't get caught is that the message format record unit tests today only test individual records not the composite record. BlobAll deserialization happens in the router tests but because encryption is not enabled this path never gets exercised.

I have now added a test as part of `MessageFormatInputStreamTest` that caught this, and I have fixed the code.